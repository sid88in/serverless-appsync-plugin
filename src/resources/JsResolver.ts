import { IntrinsicFunction } from '../types/cloudFormation';
import fs from 'fs';
import { Substitutions } from '../types/plugin';
import { Api } from './Api';
import { buildSync } from 'esbuild';

type JsResolverConfig = {
  path: string;
  substitutions?: Substitutions;
};

export class JsResolver {
  constructor(private api: Api, private config: JsResolverConfig) {}

  compile(): string | IntrinsicFunction {
    if (!fs.existsSync(this.config.path)) {
      throw new this.api.plugin.serverless.classes.Error(
        `The resolver handler file '${this.config.path}' does not exist`,
      );
    }

    return this.processTemplateSubstitutions(this.getResolverContent());
  }

  getResolverContent(): string {
    if (this.api.config.esbuild === false) {
      return fs.readFileSync(this.config.path, 'utf8');
    }

    // process with esbuild
    // this will:
    // - Bundle the code into one file if necessary
    // - Transpile typescript to javascript if necessary
    const buildResult = buildSync({
      target: 'esnext',
      sourcemap: 'inline',
      sourcesContent: false,
      treeShaking: true,
      // custom config overrides
      ...this.api.config.esbuild,
      // These options are required and can't be changed
      platform: 'node',
      format: 'esm',
      entryPoints: [this.config.path],
      bundle: true,
      write: false,
      external: ['@aws-appsync/utils'],
    });

    if (buildResult.errors.length > 0) {
      throw new this.api.plugin.serverless.classes.Error(
        `Failed to compile resolver handler file '${this.config.path}': ${buildResult.errors[0].text}`,
      );
    }

    if (buildResult.outputFiles.length === 0) {
      throw new this.api.plugin.serverless.classes.Error(
        `Failed to compile resolver handler file '${this.config.path}': No output files`,
      );
    }

    return buildResult.outputFiles[0].text;
  }

  processTemplateSubstitutions(template: string): string | IntrinsicFunction {
    const substitutions = {
      ...this.api.config.substitutions,
      ...this.config.substitutions,
    };
    const availableVariables = Object.keys(substitutions);
    const templateVariables: string[] = [];
    let searchResult;
    const variableSyntax = RegExp(/#([\w\d-_]+)#/g);
    while ((searchResult = variableSyntax.exec(template)) !== null) {
      templateVariables.push(searchResult[1]);
    }

    const replacements = availableVariables
      .filter((value) => templateVariables.includes(value))
      .filter((value, index, array) => array.indexOf(value) === index)
      .reduce(
        (accum, value) =>
          Object.assign(accum, { [value]: substitutions[value] }),
        {},
      );

    // if there are substitutions for this template then add fn:sub
    if (Object.keys(replacements).length > 0) {
      return this.substituteGlobalTemplateVariables(template, replacements);
    }

    return template;
  }

  /**
   * Creates Fn::Join object from given template where all given substitutions
   * are wrapped in Fn::Sub objects. This enables template to have also
   * characters that are not only alphanumeric, underscores, periods, and colons.
   *
   * @param {*} template
   * @param {*} substitutions
   */
  substituteGlobalTemplateVariables(
    template: string,
    substitutions: Substitutions,
  ): IntrinsicFunction {
    const variables = Object.keys(substitutions).join('|');
    const regex = new RegExp(`\\#(${variables})#`, 'g');
    const substituteTemplate = template.replace(regex, '|||$1|||');

    const templateJoin = substituteTemplate
      .split('|||')
      .filter((part) => part !== '');
    const parts: (string | IntrinsicFunction)[] = [];
    for (let i = 0; i < templateJoin.length; i += 1) {
      if (templateJoin[i] in substitutions) {
        const subs = { [templateJoin[i]]: substitutions[templateJoin[i]] };
        parts[i] = { 'Fn::Sub': [`\${${templateJoin[i]}}`, subs] };
      } else {
        parts[i] = templateJoin[i];
      }
    }
    return { 'Fn::Join': ['', parts] };
  }
}
