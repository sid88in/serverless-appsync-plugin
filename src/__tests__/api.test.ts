import { Api } from '../resources/Api';
import * as given from './given';

// 2020-12-09T16:24:22+00:00
jest.spyOn(Date, 'now').mockImplementation(() => 1607531062000);

const plugin = given.plugin();

describe('Api', () => {
  describe('compileEndpoint', () => {
    it('should compile the Api Resource', () => {
      const api = new Api(given.appSyncConfig(), plugin);
      expect(api.compileEndpoint()).toMatchInlineSnapshot(`
        Object {
          "GraphQlApi": Object {
            "Properties": Object {
              "AuthenticationType": "API_KEY",
              "EnvironmentVariables": undefined,
              "Name": "MyApi",
              "Tags": Array [
                Object {
                  "Key": "stage",
                  "Value": "Dev",
                },
              ],
              "XrayEnabled": false,
            },
            "Type": "AWS::AppSync::GraphQLApi",
          },
        }
      `);
    });

    it('should compile the Api Resource for a private endpoint', () => {
      const api = new Api(
        given.appSyncConfig({
          visibility: 'PRIVATE',
        }),
        plugin,
      );
      expect(api.compileEndpoint()).toMatchInlineSnapshot(`
        Object {
          "GraphQlApi": Object {
            "Properties": Object {
              "AuthenticationType": "API_KEY",
              "EnvironmentVariables": undefined,
              "Name": "MyApi",
              "Tags": Array [
                Object {
                  "Key": "stage",
                  "Value": "Dev",
                },
              ],
              "Visibility": "PRIVATE",
              "XrayEnabled": false,
            },
            "Type": "AWS::AppSync::GraphQLApi",
          },
        }
      `);
    });

    it('should compile the Api Resource with config', () => {
      const api = new Api(
        given.appSyncConfig({
          introspection: false,
          queryDepthLimit: 10,
          resolverCountLimit: 20,
        }),
        plugin,
      );
      expect(api.compileEndpoint()).toMatchInlineSnapshot(`
        Object {
          "GraphQlApi": Object {
            "Properties": Object {
              "AuthenticationType": "API_KEY",
              "EnvironmentVariables": undefined,
              "IntrospectionConfig": "DISABLED",
              "Name": "MyApi",
              "QueryDepthLimit": 10,
              "ResolverCountLimit": 20,
              "Tags": Array [
                Object {
                  "Key": "stage",
                  "Value": "Dev",
                },
              ],
              "XrayEnabled": false,
            },
            "Type": "AWS::AppSync::GraphQLApi",
          },
        }
      `);
    });

    it('should compile the Api Resource with Environments', () => {
      const api = new Api(
        given.appSyncConfig({
          environment: {
            TABLE_NAME: 'MyTable',
            OTHER_TABLE: {
              Ref: 'OtherTable',
            },
          },
        }),
        plugin,
      );
      expect(api.compileEndpoint()).toMatchInlineSnapshot(`
        Object {
          "GraphQlApi": Object {
            "Properties": Object {
              "AuthenticationType": "API_KEY",
              "EnvironmentVariables": Object {
                "OTHER_TABLE": Object {
                  "Ref": "OtherTable",
                },
                "TABLE_NAME": "MyTable",
              },
              "Name": "MyApi",
              "Tags": Array [
                Object {
                  "Key": "stage",
                  "Value": "Dev",
                },
              ],
              "XrayEnabled": false,
            },
            "Type": "AWS::AppSync::GraphQLApi",
          },
        }
      `);
    });

    it('should compile the Api Resource with logs enabled', () => {
      const api = new Api(
        given.appSyncConfig({
          logging: {
            level: 'ERROR',
            excludeVerboseContent: false,
            retentionInDays: 14,
          },
        }),
        plugin,
      );
      expect(api.compileEndpoint()).toMatchInlineSnapshot(`
        Object {
          "GraphQlApi": Object {
            "Properties": Object {
              "AuthenticationType": "API_KEY",
              "EnvironmentVariables": undefined,
              "LogConfig": Object {
                "CloudWatchLogsRoleArn": Object {
                  "Fn::GetAtt": Array [
                    "GraphQlApiLogGroupRole",
                    "Arn",
                  ],
                },
                "ExcludeVerboseContent": false,
                "FieldLogLevel": "ERROR",
              },
              "Name": "MyApi",
              "Tags": Array [
                Object {
                  "Key": "stage",
                  "Value": "Dev",
                },
              ],
              "XrayEnabled": false,
            },
            "Type": "AWS::AppSync::GraphQLApi",
          },
        }
      `);
    });

    it('should compile the Api Resource with additional auths', () => {
      const api = new Api(
        given.appSyncConfig({
          authentication: {
            type: 'AMAZON_COGNITO_USER_POOLS',
            config: {
              userPoolId: 'pool123',
              awsRegion: 'us-east-1',
              appIdClientRegex: '[a-z]',
            },
          },
          additionalAuthentications: [
            {
              type: 'AMAZON_COGNITO_USER_POOLS',
              config: {
                userPoolId: 'pool123',
                awsRegion: 'us-east-1',
                appIdClientRegex: '[a-z]',
              },
            },
            {
              type: 'AWS_IAM',
            },
            {
              type: 'OPENID_CONNECT',
              config: {
                issuer: 'https://auth.example.com',
                clientId: '333746dd-06fc-44df-bce2-5ff108724044',
                iatTTL: 3600,
                authTTL: 60,
              },
            },
            {
              type: 'AWS_LAMBDA',
              config: {
                functionName: 'authFunction',
                identityValidationExpression: 'customm-*',
                authorizerResultTtlInSeconds: 300,
              },
            },
          ],
        }),
        plugin,
      );
      expect(api.compileEndpoint()).toMatchInlineSnapshot(`
        Object {
          "GraphQlApi": Object {
            "Properties": Object {
              "AdditionalAuthenticationProviders": Array [
                Object {
                  "AuthenticationType": "AMAZON_COGNITO_USER_POOLS",
                  "UserPoolConfig": Object {
                    "AppIdClientRegex": "[a-z]",
                    "AwsRegion": "us-east-1",
                    "UserPoolId": "pool123",
                  },
                },
                Object {
                  "AuthenticationType": "AWS_IAM",
                },
                Object {
                  "AuthenticationType": "OPENID_CONNECT",
                  "OpenIDConnectConfig": Object {
                    "AuthTTL": 60,
                    "ClientId": "333746dd-06fc-44df-bce2-5ff108724044",
                    "IatTTL": 3600,
                    "Issuer": "https://auth.example.com",
                  },
                },
                Object {
                  "AuthenticationType": "AWS_LAMBDA",
                  "LambdaAuthorizerConfig": Object {
                    "AuthorizerResultTtlInSeconds": 300,
                    "AuthorizerUri": Object {
                      "Fn::GetAtt": Array [
                        "AuthFunctionLambdaFunction",
                        "Arn",
                      ],
                    },
                    "IdentityValidationExpression": "customm-*",
                  },
                },
              ],
              "AuthenticationType": "AMAZON_COGNITO_USER_POOLS",
              "EnvironmentVariables": undefined,
              "Name": "MyApi",
              "Tags": Array [
                Object {
                  "Key": "stage",
                  "Value": "Dev",
                },
              ],
              "UserPoolConfig": Object {
                "AppIdClientRegex": "[a-z]",
                "AwsRegion": "us-east-1",
                "DefaultAction": "ALLOW",
                "UserPoolId": "pool123",
              },
              "XrayEnabled": false,
            },
            "Type": "AWS::AppSync::GraphQLApi",
          },
        }
      `);
    });

    it('should compile the Api Resource with embedded authorizer Lambda function', () => {
      const api = new Api(
        given.appSyncConfig({
          authentication: {
            type: 'AWS_LAMBDA',
            config: {
              function: {
                handler: 'index.handler',
              },
            },
          },
        }),
        plugin,
      );
      expect(api.compileEndpoint()).toMatchSnapshot();
      expect(api.functions).toMatchSnapshot();
    });

    it('should compile the Api Resource with embedded additional authorizer Lambda function', () => {
      const api = new Api(
        given.appSyncConfig({
          additionalAuthentications: [
            {
              type: 'AWS_LAMBDA',
              config: {
                function: {
                  handler: 'index.handler',
                },
              },
            },
          ],
        }),
        plugin,
      );
      expect(api.compileEndpoint()).toMatchSnapshot();
      expect(api.functions).toMatchSnapshot();
    });
  });

  describe('Logs', () => {
    it('should not compile CloudWatch Resources when logging not configured', () => {
      const api = new Api(given.appSyncConfig(), plugin);
      expect(api.compileCloudWatchLogGroup()).toMatchInlineSnapshot(
        `Object {}`,
      );
    });

    it('should not compile CloudWatch Resources when logging is disabled', () => {
      const api = new Api(
        given.appSyncConfig({
          logging: {
            level: 'ERROR',
            retentionInDays: 14,
            enabled: false,
          },
        }),
        plugin,
      );
      expect(api.compileCloudWatchLogGroup()).toMatchInlineSnapshot(
        `Object {}`,
      );
    });

    it('should compile CloudWatch Resources when enaabled', () => {
      const api = new Api(
        given.appSyncConfig({
          logging: {
            level: 'ERROR',
            retentionInDays: 14,
          },
        }),
        plugin,
      );
      expect(api.compileCloudWatchLogGroup()).toMatchInlineSnapshot(`
        Object {
          "GraphQlApiLogGroup": Object {
            "Properties": Object {
              "LogGroupName": Object {
                "Fn::Join": Array [
                  "/",
                  Array [
                    "/aws/appsync/apis",
                    Object {
                      "Fn::GetAtt": Array [
                        "GraphQlApi",
                        "ApiId",
                      ],
                    },
                  ],
                ],
              },
              "RetentionInDays": 14,
            },
            "Type": "AWS::Logs::LogGroup",
          },
          "GraphQlApiLogGroupPolicy": Object {
            "Properties": Object {
              "PolicyDocument": Object {
                "Statement": Array [
                  Object {
                    "Action": Array [
                      "logs:CreateLogGroup",
                      "logs:CreateLogStream",
                      "logs:PutLogEvents",
                    ],
                    "Effect": "Allow",
                    "Resource": Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "GraphQlApiLogGroup",
                          "Arn",
                        ],
                      },
                    ],
                  },
                ],
                "Version": "2012-10-17",
              },
              "PolicyName": "GraphQlApiLogGroupPolicy",
              "Roles": Array [
                Object {
                  "Ref": "GraphQlApiLogGroupRole",
                },
              ],
            },
            "Type": "AWS::IAM::Policy",
          },
          "GraphQlApiLogGroupRole": Object {
            "Properties": Object {
              "AssumeRolePolicyDocument": Object {
                "Statement": Array [
                  Object {
                    "Action": Array [
                      "sts:AssumeRole",
                    ],
                    "Effect": "Allow",
                    "Principal": Object {
                      "Service": Array [
                        "appsync.amazonaws.com",
                      ],
                    },
                  },
                ],
                "Version": "2012-10-17",
              },
            },
            "Type": "AWS::IAM::Role",
          },
        }
      `);
    });
  });

  describe('apiKeys', () => {
    const api = new Api(given.appSyncConfig(), plugin);

    it('should generate an api key with sliding window expiration in numeric hours', () => {
      expect(
        api.compileApiKey({
          name: 'Default',
          description: 'Default Key',
          expiresAfter: 24,
        }),
      ).toMatchInlineSnapshot(`
        Object {
          "GraphQlApiDefault": Object {
            "Properties": Object {
              "ApiId": Object {
                "Fn::GetAtt": Array [
                  "GraphQlApi",
                  "ApiId",
                ],
              },
              "ApiKeyId": undefined,
              "Description": "Default Key",
              "Expires": 1607619600,
            },
            "Type": "AWS::AppSync::ApiKey",
          },
        }
      `);
    });

    it('should generate an api key with sliding window expiration in string hours', () => {
      expect(
        api.compileApiKey({
          name: 'Default',
          description: 'Default Key',
          expiresAfter: '24',
        }),
      ).toMatchInlineSnapshot(`
        Object {
          "GraphQlApiDefault": Object {
            "Properties": Object {
              "ApiId": Object {
                "Fn::GetAtt": Array [
                  "GraphQlApi",
                  "ApiId",
                ],
              },
              "ApiKeyId": undefined,
              "Description": "Default Key",
              "Expires": 1607619600,
            },
            "Type": "AWS::AppSync::ApiKey",
          },
        }
      `);
    });

    it('should generate an api key with sliding window expiration in duration', () => {
      expect(
        api.compileApiKey({
          name: 'Default',
          description: 'Default Key',
          expiresAfter: '30d',
        }),
      ).toMatchInlineSnapshot(`
        Object {
          "GraphQlApiDefault": Object {
            "Properties": Object {
              "ApiId": Object {
                "Fn::GetAtt": Array [
                  "GraphQlApi",
                  "ApiId",
                ],
              },
              "ApiKeyId": undefined,
              "Description": "Default Key",
              "Expires": 1610121600,
            },
            "Type": "AWS::AppSync::ApiKey",
          },
        }
      `);
    });

    it('should generate an api key with explicit expiresAt', () => {
      expect(
        api.compileApiKey({
          name: 'Default',
          description: 'Default Key',
          expiresAt: '2022-12-31T22:00:00+00:00',
        }),
      ).toMatchInlineSnapshot(`
        Object {
          "GraphQlApiDefault": Object {
            "Properties": Object {
              "ApiId": Object {
                "Fn::GetAtt": Array [
                  "GraphQlApi",
                  "ApiId",
                ],
              },
              "ApiKeyId": undefined,
              "Description": "Default Key",
              "Expires": 1672524000,
            },
            "Type": "AWS::AppSync::ApiKey",
          },
        }
      `);
    });

    it('should generate an api key with default expiry', () => {
      expect(
        api.compileApiKey({
          name: 'Default',
          description: 'Default Key',
        }),
      ).toMatchInlineSnapshot(`
        Object {
          "GraphQlApiDefault": Object {
            "Properties": Object {
              "ApiId": Object {
                "Fn::GetAtt": Array [
                  "GraphQlApi",
                  "ApiId",
                ],
              },
              "ApiKeyId": undefined,
              "Description": "Default Key",
              "Expires": 1639065600,
            },
            "Type": "AWS::AppSync::ApiKey",
          },
        }
      `);
    });
  });

  describe('LambdaAuthorizer', () => {
    it('should not generate the Lambda Authorizer Resources', () => {
      const api = new Api(
        given.appSyncConfig({
          authentication: {
            type: 'API_KEY',
          },
        }),
        plugin,
      );
      expect(api.compileLambdaAuthorizerPermission()).toMatchInlineSnapshot(
        `Object {}`,
      );
    });

    it('should generate the Lambda Authorizer Resources from basic auth', () => {
      const api = new Api(
        given.appSyncConfig({
          authentication: {
            type: 'AWS_LAMBDA',
            config: {
              functionArn: 'arn:',
            },
          },
        }),
        plugin,
      );
      expect(api.compileLambdaAuthorizerPermission()).toMatchInlineSnapshot(`
        Object {
          "LambdaAuthorizerPermission": Object {
            "Properties": Object {
              "Action": "lambda:InvokeFunction",
              "FunctionName": "arn:",
              "Principal": "appsync.amazonaws.com",
              "SourceArn": Object {
                "Ref": "GraphQlApi",
              },
            },
            "Type": "AWS::Lambda::Permission",
          },
        }
      `);
    });

    it('should generate the Lambda Authorizer Resources from additional auth', () => {
      const api = new Api(
        given.appSyncConfig({
          additionalAuthentications: [
            {
              type: 'AWS_LAMBDA',
              config: {
                functionArn: 'arn:',
              },
            },
          ],
        }),
        plugin,
      );
      expect(api.compileLambdaAuthorizerPermission()).toMatchInlineSnapshot(`
        Object {
          "LambdaAuthorizerPermission": Object {
            "Properties": Object {
              "Action": "lambda:InvokeFunction",
              "FunctionName": "arn:",
              "Principal": "appsync.amazonaws.com",
              "SourceArn": Object {
                "Ref": "GraphQlApi",
              },
            },
            "Type": "AWS::Lambda::Permission",
          },
        }
      `);
    });
  });
});

describe('Caching', () => {
  it('should not generate Resources when not configured', () => {
    const api = new Api(given.appSyncConfig({ caching: undefined }), plugin);
    expect(api.compileCachingResources()).toEqual({});
  });

  it('should not generate Resources when disabled', () => {
    const api = new Api(
      given.appSyncConfig({
        caching: { enabled: false, behavior: 'FULL_REQUEST_CACHING' },
      }),
      plugin,
    );
    expect(api.compileCachingResources()).toEqual({});
  });

  it('should generate Resources with defaults', () => {
    const api = new Api(
      given.appSyncConfig({
        caching: {
          behavior: 'FULL_REQUEST_CACHING',
        },
      }),
      plugin,
    );
    expect(api.compileCachingResources()).toMatchInlineSnapshot(`
        Object {
          "GraphQlCaching": Object {
            "Properties": Object {
              "ApiCachingBehavior": "FULL_REQUEST_CACHING",
              "ApiId": Object {
                "Fn::GetAtt": Array [
                  "GraphQlApi",
                  "ApiId",
                ],
              },
              "AtRestEncryptionEnabled": false,
              "TransitEncryptionEnabled": false,
              "Ttl": 3600,
              "Type": "T2_SMALL",
            },
            "Type": "AWS::AppSync::ApiCache",
          },
        }
      `);
  });

  it('should generate Resources with custom Config', () => {
    const api = new Api(
      given.appSyncConfig({
        caching: {
          behavior: 'FULL_REQUEST_CACHING',
          atRestEncryption: true,
          transitEncryption: true,
          ttl: 500,
          type: 'T2_MEDIUM',
        },
      }),
      plugin,
    );
    expect(api.compileCachingResources()).toMatchInlineSnapshot(`
        Object {
          "GraphQlCaching": Object {
            "Properties": Object {
              "ApiCachingBehavior": "FULL_REQUEST_CACHING",
              "ApiId": Object {
                "Fn::GetAtt": Array [
                  "GraphQlApi",
                  "ApiId",
                ],
              },
              "AtRestEncryptionEnabled": true,
              "TransitEncryptionEnabled": true,
              "Ttl": 500,
              "Type": "T2_MEDIUM",
            },
            "Type": "AWS::AppSync::ApiCache",
          },
        }
      `);
  });
});

describe('Domains', () => {
  it('should not generate domain resources when not configured', () => {
    const api = new Api(given.appSyncConfig({ domain: undefined }), plugin);
    expect(api.compileCustomDomain()).toMatchInlineSnapshot(`Object {}`);
  });

  it('should not generate domain resources when disabled', () => {
    const api = new Api(
      given.appSyncConfig({
        domain: {
          enabled: false,
          name: 'api.example.com',
          certificateArn:
            'arn:aws:acm:us-east-1:1234567890:certificate/e4b6e9be-1aa7-458d-880e-069622e5be52',
        },
      }),
      plugin,
    );
    expect(api.compileCustomDomain()).toMatchInlineSnapshot(`Object {}`);
  });

  it('should generate domain resources', () => {
    const api = new Api(
      given.appSyncConfig({
        domain: {
          name: 'api.example.com',
          hostedZoneId: `Z111111QQQQQQQ`,
        },
      }),
      plugin,
    );
    expect(api.compileCustomDomain()).toMatchSnapshot();
  });

  it('should generate domain resources with custom certificate ARN', () => {
    const api = new Api(
      given.appSyncConfig({
        domain: {
          name: 'api.example.com',
          certificateArn:
            'arn:aws:acm:us-east-1:1234567890:certificate/e4b6e9be-1aa7-458d-880e-069622e5be52',
        },
      }),
      plugin,
    );
    expect(api.compileCustomDomain()).toMatchSnapshot();
  });

  it('should not generate Route53 Record when disabled', () => {
    const api = new Api(
      given.appSyncConfig({
        domain: {
          name: 'api.example.com',
          certificateArn:
            'arn:aws:acm:us-east-1:1234567890:certificate/e4b6e9be-1aa7-458d-880e-069622e5be52',
          route53: false,
        },
      }),
      plugin,
    );
    expect(
      api.compileCustomDomain().GraphQlDomainRoute53Record,
    ).toBeUndefined();
  });

  it('should generate domain resources with custom hostedZoneId', () => {
    const api = new Api(
      given.appSyncConfig({
        domain: {
          name: 'api.example.com',
          certificateArn:
            'arn:aws:acm:us-east-1:1234567890:certificate/e4b6e9be-1aa7-458d-880e-069622e5be52',
          hostedZoneId: 'Z111111QQQQQQQ',
          route53: true,
        },
      }),
      plugin,
    );
    expect(api.compileCustomDomain()).toMatchSnapshot();
  });

  it('should generate domain resources with custom hostedZoneName', () => {
    const api = new Api(
      given.appSyncConfig({
        domain: {
          name: 'foo.api.example.com',
          certificateArn:
            'arn:aws:acm:us-east-1:1234567890:certificate/e4b6e9be-1aa7-458d-880e-069622e5be52',
          hostedZoneName: 'example.com.',
          route53: true,
        },
      }),
      plugin,
    );
    expect(api.compileCustomDomain()).toMatchSnapshot();
  });
});
