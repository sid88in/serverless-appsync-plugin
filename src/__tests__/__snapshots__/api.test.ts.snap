// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Api compileEndpoint should compile the Api Resource with embedded additional authorizer Lambda function 1`] = `
Object {
  "GraphQlApi": Object {
    "Properties": Object {
      "AdditionalAuthenticationProviders": Array [
        Object {
          "AuthenticationType": "AWS_LAMBDA",
          "LambdaAuthorizerConfig": Object {
            "AuthorizerResultTtlInSeconds": undefined,
            "AuthorizerUri": Object {
              "Fn::GetAtt": Array [
                "MyApiAuthorizerLambdaFunction",
                "Arn",
              ],
            },
            "IdentityValidationExpression": undefined,
          },
        },
      ],
      "AuthenticationType": "API_KEY",
      "EnvironmentVariables": undefined,
      "Name": "MyApi",
      "Tags": Array [
        Object {
          "Key": "stage",
          "Value": "Dev",
        },
      ],
      "XrayEnabled": false,
    },
    "Type": "AWS::AppSync::GraphQLApi",
  },
}
`;

exports[`Api compileEndpoint should compile the Api Resource with embedded additional authorizer Lambda function 2`] = `
Object {
  "MyApiAuthorizer": Object {
    "handler": "index.handler",
  },
}
`;

exports[`Api compileEndpoint should compile the Api Resource with embedded authorizer Lambda function 1`] = `
Object {
  "GraphQlApi": Object {
    "Properties": Object {
      "AuthenticationType": "AWS_LAMBDA",
      "EnvironmentVariables": undefined,
      "LambdaAuthorizerConfig": Object {
        "AuthorizerResultTtlInSeconds": undefined,
        "AuthorizerUri": Object {
          "Fn::GetAtt": Array [
            "MyApiAuthorizerLambdaFunction",
            "Arn",
          ],
        },
        "IdentityValidationExpression": undefined,
      },
      "Name": "MyApi",
      "Tags": Array [
        Object {
          "Key": "stage",
          "Value": "Dev",
        },
      ],
      "XrayEnabled": false,
    },
    "Type": "AWS::AppSync::GraphQLApi",
  },
}
`;

exports[`Api compileEndpoint should compile the Api Resource with embedded authorizer Lambda function 2`] = `
Object {
  "MyApiAuthorizer": Object {
    "handler": "index.handler",
  },
}
`;

exports[`Domains should generate domain resources 1`] = `
Object {
  "GraphQlDomainAssociation": Object {
    "DeletionPolicy": "Delete",
    "DependsOn": Array [
      "GraphQlDomainName",
    ],
    "Properties": Object {
      "ApiId": Object {
        "Fn::GetAtt": Array [
          "GraphQlApi",
          "ApiId",
        ],
      },
      "DomainName": "api.example.com",
    },
    "Type": "AWS::AppSync::DomainNameApiAssociation",
  },
  "GraphQlDomainCertificate": Object {
    "DeletionPolicy": "Delete",
    "Properties": Object {
      "DomainName": "api.example.com",
      "DomainValidationOptions": Array [
        Object {
          "DomainName": "api.example.com",
          "HostedZoneId": "Z111111QQQQQQQ",
        },
      ],
      "ValidationMethod": "DNS",
    },
    "Type": "AWS::CertificateManager::Certificate",
  },
  "GraphQlDomainName": Object {
    "DeletionPolicy": "Delete",
    "Properties": Object {
      "CertificateArn": Object {
        "Ref": "GraphQlDomainCertificate",
      },
      "DomainName": "api.example.com",
    },
    "Type": "AWS::AppSync::DomainName",
  },
  "GraphQlDomainRoute53Record": Object {
    "DeletionPolicy": "Delete",
    "Properties": Object {
      "AliasTarget": Object {
        "DNSName": Object {
          "Fn::GetAtt": Array [
            "GraphQlDomainName",
            "AppSyncDomainName",
          ],
        },
        "EvaluateTargetHealth": false,
        "HostedZoneId": Object {
          "Fn::GetAtt": Array [
            "GraphQlDomainName",
            "HostedZoneId",
          ],
        },
      },
      "HostedZoneId": "Z111111QQQQQQQ",
      "Name": "api.example.com",
      "Type": "A",
    },
    "Type": "AWS::Route53::RecordSet",
  },
}
`;

exports[`Domains should generate domain resources with custom certificate ARN 1`] = `
Object {
  "GraphQlDomainAssociation": Object {
    "DeletionPolicy": "Delete",
    "DependsOn": Array [
      "GraphQlDomainName",
    ],
    "Properties": Object {
      "ApiId": Object {
        "Fn::GetAtt": Array [
          "GraphQlApi",
          "ApiId",
        ],
      },
      "DomainName": "api.example.com",
    },
    "Type": "AWS::AppSync::DomainNameApiAssociation",
  },
  "GraphQlDomainName": Object {
    "DeletionPolicy": "Delete",
    "Properties": Object {
      "CertificateArn": "arn:aws:acm:us-east-1:1234567890:certificate/e4b6e9be-1aa7-458d-880e-069622e5be52",
      "DomainName": "api.example.com",
    },
    "Type": "AWS::AppSync::DomainName",
  },
  "GraphQlDomainRoute53Record": Object {
    "DeletionPolicy": "Delete",
    "Properties": Object {
      "AliasTarget": Object {
        "DNSName": Object {
          "Fn::GetAtt": Array [
            "GraphQlDomainName",
            "AppSyncDomainName",
          ],
        },
        "EvaluateTargetHealth": false,
        "HostedZoneId": Object {
          "Fn::GetAtt": Array [
            "GraphQlDomainName",
            "HostedZoneId",
          ],
        },
      },
      "HostedZoneName": "example.com.",
      "Name": "api.example.com",
      "Type": "A",
    },
    "Type": "AWS::Route53::RecordSet",
  },
}
`;

exports[`Domains should generate domain resources with custom hostedZoneId 1`] = `
Object {
  "GraphQlDomainAssociation": Object {
    "DeletionPolicy": "Delete",
    "DependsOn": Array [
      "GraphQlDomainName",
    ],
    "Properties": Object {
      "ApiId": Object {
        "Fn::GetAtt": Array [
          "GraphQlApi",
          "ApiId",
        ],
      },
      "DomainName": "api.example.com",
    },
    "Type": "AWS::AppSync::DomainNameApiAssociation",
  },
  "GraphQlDomainName": Object {
    "DeletionPolicy": "Delete",
    "Properties": Object {
      "CertificateArn": "arn:aws:acm:us-east-1:1234567890:certificate/e4b6e9be-1aa7-458d-880e-069622e5be52",
      "DomainName": "api.example.com",
    },
    "Type": "AWS::AppSync::DomainName",
  },
  "GraphQlDomainRoute53Record": Object {
    "DeletionPolicy": "Delete",
    "Properties": Object {
      "AliasTarget": Object {
        "DNSName": Object {
          "Fn::GetAtt": Array [
            "GraphQlDomainName",
            "AppSyncDomainName",
          ],
        },
        "EvaluateTargetHealth": false,
        "HostedZoneId": Object {
          "Fn::GetAtt": Array [
            "GraphQlDomainName",
            "HostedZoneId",
          ],
        },
      },
      "HostedZoneId": "Z111111QQQQQQQ",
      "Name": "api.example.com",
      "Type": "A",
    },
    "Type": "AWS::Route53::RecordSet",
  },
}
`;

exports[`Domains should generate domain resources with custom hostedZoneName 1`] = `
Object {
  "GraphQlDomainAssociation": Object {
    "DeletionPolicy": "Delete",
    "DependsOn": Array [
      "GraphQlDomainName",
    ],
    "Properties": Object {
      "ApiId": Object {
        "Fn::GetAtt": Array [
          "GraphQlApi",
          "ApiId",
        ],
      },
      "DomainName": "foo.api.example.com",
    },
    "Type": "AWS::AppSync::DomainNameApiAssociation",
  },
  "GraphQlDomainName": Object {
    "DeletionPolicy": "Delete",
    "Properties": Object {
      "CertificateArn": "arn:aws:acm:us-east-1:1234567890:certificate/e4b6e9be-1aa7-458d-880e-069622e5be52",
      "DomainName": "foo.api.example.com",
    },
    "Type": "AWS::AppSync::DomainName",
  },
  "GraphQlDomainRoute53Record": Object {
    "DeletionPolicy": "Delete",
    "Properties": Object {
      "AliasTarget": Object {
        "DNSName": Object {
          "Fn::GetAtt": Array [
            "GraphQlDomainName",
            "AppSyncDomainName",
          ],
        },
        "EvaluateTargetHealth": false,
        "HostedZoneId": Object {
          "Fn::GetAtt": Array [
            "GraphQlDomainName",
            "HostedZoneId",
          ],
        },
      },
      "HostedZoneName": "example.com.",
      "Name": "foo.api.example.com",
      "Type": "A",
    },
    "Type": "AWS::Route53::RecordSet",
  },
}
`;
